x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/search/SearchApplication.java

```diff
@@ -0,0 +1,261 @@
+
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+
+import static org.elasticsearch.xcontent.ConstructingObjectParser.constructorArg;
+import static org.elasticsearch.xcontent.ConstructingObjectParser.optionalConstructorArg;
+
+public class SearchApplication implements Writeable, ToXContentObject {
+
+    private final String name;
+    private final String[] indices;
+    private static final ConstructingObjectParser<SearchApplication, String> PARSER = new ConstructingObjectParser<>(
+        "search_application",
+        false,
+        (params, resourceName) -> {
+            final String name = (String) params[0];
+            // If name is provided, check that it matches the resource name. We don't want it to be updatable
+            if (name != null) {
+                if (name.equals(resourceName) == false) {
+                    throw new IllegalArgumentException(
+                        "Search Application name [" + name + "] does not match the resource name: [" + resourceName + "]"
+                    );
+                }
+            }
+            @SuppressWarnings("unchecked")
+            final String[] indices = ((List<String>) params[1]).toArray(String[]::new);
+            final String analyticsCollectionName = (String) params[2];
+            final Long maybeUpdatedAtMillis = (Long) params[3];
+            long updatedAtMillis = (maybeUpdatedAtMillis != null ? maybeUpdatedAtMillis : System.currentTimeMillis());
+
+            SearchApplication newApp = new SearchApplication(resourceName, indices, analyticsCollectionName, updatedAtMillis);
+            return newApp;
+        }
+    );
+    private final String analyticsCollectionName;
+    private final long updatedAtMillis;
+
+    public SearchApplication(StreamInput in) throws IOException {
+        this.name = in.readString();
+        this.indices = in.readStringArray();
+        this.analyticsCollectionName = in.readOptionalString();
+        this.updatedAtMillis = in.readLong();
+    }
+
+    /**
+     * Public constructor.
+     *
+     * @param name                    The name of the search application.
+     * @param indices                 The list of indices targeted by this search application.
+     * @param analyticsCollectionName The name of the associated analytics collection.
+     * @param updatedAtMillis         Last updated time in milliseconds for the search application.
+     */
+    public SearchApplication(String name, String[] indices, @Nullable String analyticsCollectionName, long updatedAtMillis) {
+        if (Strings.isNullOrEmpty(name)) {
+            throw new IllegalArgumentException("Search Application name cannot be null or blank");
+        }
+        this.name = name;
+
+        Objects.requireNonNull(indices, "Search Application indices cannot be null");
+        this.indices = indices.clone();
```

Comment:
Don't use clone. I know its used elsewhere, but it is much preferred to use `Arrays.copyOf`.
```suggestion
        this.indices = Arrays.copyOf(indices, indices.length);
```